[ICE]: delegation: unreachable

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_hir_analysis/src/delegation.rs:229:45: 'internal error: entered unreachable code'', 'thread 'rustc' ($TID) panicked at compiler/rustc_hir_analysis/src/delegation.rs:229:45: 'internal error: entered unreachable code''
File: /tmp/im/7/a.rs
-->

auto-reduced (treereduce-rust):
````rust
#![feature(min_generic_const_args)]

impl
    S<
        {
            fn foo() {}
            reuse foo::<> as bar;
            reuse bar;
        },
    >
{
}

````

original:
````rust
#![feature(min_generic_const_args)]
#![feature(fn_delegation)]

pub struct S<const N: usize>;
impl
    S<
        {
            fn foo() {}
            reuse foo::<> as bar;
            reuse bar;
        },
    >
{
}
fn main() {}

````

Version information
````
rustc 1.97.0-nightly (12f35ad39 2026-04-14)
binary: rustc
commit-hash: 12f35ad39ed3e39df4d953c46d4f6cc6c82adc96
commit-date: 2026-04-14
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/12f35ad39ed3e39df4d953c46d4f6cc6c82adc96/compiler/rustc_hir_analysis/src/delegation.rs#L223-L235

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0428]: the name `bar` is defined multiple times
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:8:13
  |
7 |             reuse foo::<> as bar;
  |             --------------------- previous definition of the value `bar` here
8 |             reuse bar;
  |             ^^^^^^^^^^ `bar` redefined here
  |
  = note: `bar` must be defined only once in the value namespace of this block

error[E0425]: cannot find type `S` in this scope
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:4:5
  |
4 |     S<
  |     ^ not found in this scope

error[E0658]: functions delegation is not yet fully implemented
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:7:13
  |
7 |             reuse foo::<> as bar;
  |             ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-14; consider upgrading it if it is out of date

error[E0658]: functions delegation is not yet fully implemented
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:8:13
  |
8 |             reuse bar;
  |             ^^^^^^^^^^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-14; consider upgrading it if it is out of date

warning: the feature `min_generic_const_args` is incomplete and may not be safe to use and/or cause compiler crashes
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:1:12
  |
1 | #![feature(min_generic_const_args)]
  |            ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #132980 <https://github.com/rust-lang/rust/issues/132980> for more information
  = note: `#[warn(incomplete_features)]` on by default

error: complex const arguments must be placed inside of a `const` block
 --> /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:5:9
  |
5 | /         {
6 | |             fn foo() {}
7 | |             reuse foo::<> as bar;
8 | |             reuse bar;
9 | |         },
  | |_________^


thread 'rustc' (978165) panicked at /rustc-dev/12f35ad39ed3e39df4d953c46d4f6cc6c82adc96/compiler/rustc_hir_analysis/src/delegation.rs:229:45:
internal error: entered unreachable code
stack backtrace:
   0:     0x7f8a8ce217eb - <<std[f4db05f3516471ac]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[7428369f851be300]::fmt::Display>::fmt
   1:     0x7f8a8d421748 - core[7428369f851be300]::fmt::write
   2:     0x7f8a8ce38ab6 - <std[f4db05f3516471ac]::sys::stdio::unix::Stderr as std[f4db05f3516471ac]::io::Write>::write_fmt
   3:     0x7f8a8cdf7ae8 - std[f4db05f3516471ac]::panicking::default_hook::{closure#0}
   4:     0x7f8a8ce14e53 - std[f4db05f3516471ac]::panicking::default_hook
   5:     0x7f8a8bdec29c - std[f4db05f3516471ac]::panicking::update_hook::<alloc[35d6e4a9b7d8a159]::boxed::Box<rustc_driver_impl[c394bdd49c3d550c]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f8a8ce15132 - std[f4db05f3516471ac]::panicking::panic_with_hook
   7:     0x7f8a8cdf7bda - std[f4db05f3516471ac]::panicking::panic_handler::{closure#0}
   8:     0x7f8a8cdee8e9 - std[f4db05f3516471ac]::sys::backtrace::__rust_end_short_backtrace::<std[f4db05f3516471ac]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f8a8cdf960d - __rustc[11f1a118111b43c5]::rust_begin_unwind
  10:     0x7f8a8994e43c - core[7428369f851be300]::panicking::panic_fmt
  11:     0x7f8a8ab2fb52 - core[7428369f851be300]::panicking::panic
  12:     0x7f8a8bf898a0 - rustc_hir_analysis[6c0c0d88845e98ad]::delegation::get_delegation_parent_args_count_without_self
  13:     0x7f8a8bf85b38 - rustc_hir_analysis[6c0c0d88845e98ad]::delegation::create_folder_and_args
  14:     0x7f8a8bf87301 - rustc_hir_analysis[6c0c0d88845e98ad]::delegation::inherit_sig_for_delegation_item
  15:     0x7f8a8c8c0be2 - rustc_query_impl[66bcb677d88b0cff]::query_impl::inherit_sig_for_delegation_item::invoke_provider_fn::__rust_begin_short_backtrace
  16:     0x7f8a8dd6ad9d - rustc_query_impl[66bcb677d88b0cff]::execution::try_execute_query::<rustc_data_structures[79cea45f287b0693]::vec_cache::VecCache<rustc_span[11e87916d9377e75]::def_id::LocalDefId, rustc_middle[48c27bf4c0e69198]::query::erase::ErasedData<[u8; 16usize]>, rustc_middle[48c27bf4c0e69198]::dep_graph::graph::DepNodeIndex>, false>
  17:     0x7f8a8c8c0c4a - rustc_query_impl[66bcb677d88b0cff]::query_impl::inherit_sig_for_delegation_item::execute_query_non_incr::__rust_end_short_backtrace
  18:     0x7f8a8db99fc4 - <dyn rustc_hir_analysis[6c0c0d88845e98ad]::hir_ty_lowering::HirTyLowerer>::lower_ty
  19:     0x7f8a8dbabe96 - <dyn rustc_hir_analysis[6c0c0d88845e98ad]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
  20:     0x7f8a8dbab683 - rustc_hir_analysis[6c0c0d88845e98ad]::collect::lower_fn_sig_recovering_infer_ret_ty
  21:     0x7f8a8d6d8258 - rustc_hir_analysis[6c0c0d88845e98ad]::collect::fn_sig
  22:     0x7f8a8d6d7926 - rustc_query_impl[66bcb677d88b0cff]::query_impl::fn_sig::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7f8a8d6dbe9b - rustc_query_impl[66bcb677d88b0cff]::execution::try_execute_query::<rustc_middle[48c27bf4c0e69198]::query::caches::DefIdCache<rustc_middle[48c27bf4c0e69198]::query::erase::ErasedData<[u8; 24usize]>>, false>
  24:     0x7f8a8d6db465 - rustc_query_impl[66bcb677d88b0cff]::query_impl::fn_sig::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7f8a8b87d69c - <rustc_ast_lowering[1289cac3f79600f4]::LoweringContext<rustc_ast_lowering[1289cac3f79600f4]::ResolverDelayedAstLowering>>::lower_delegation
  26:     0x7f8a8b899df4 - <rustc_ast_lowering[1289cac3f79600f4]::LoweringContext<rustc_ast_lowering[1289cac3f79600f4]::ResolverDelayedAstLowering>>::lower_item_kind
  27:     0x7f8a8b8c6def - <rustc_ast_lowering[1289cac3f79600f4]::item::ItemLowerer<rustc_ast_lowering[1289cac3f79600f4]::ResolverDelayedAstLowering>>::lower_node
  28:     0x7f8a8b87bbc9 - rustc_ast_lowering[1289cac3f79600f4]::lower_delayed_owner
  29:     0x7f8a8dec1e62 - rustc_query_impl[66bcb677d88b0cff]::execution::try_execute_query::<rustc_data_structures[79cea45f287b0693]::vec_cache::VecCache<rustc_span[11e87916d9377e75]::def_id::LocalDefId, rustc_middle[48c27bf4c0e69198]::query::erase::ErasedData<[u8; 0usize]>, rustc_middle[48c27bf4c0e69198]::dep_graph::graph::DepNodeIndex>, false>
  30:     0x7f8a8c8bb1bb - rustc_query_impl[66bcb677d88b0cff]::query_impl::lower_delayed_owner::execute_query_non_incr::__rust_end_short_backtrace
  31:     0x7f8a8e5ed609 - rustc_middle[48c27bf4c0e69198]::hir::map::hir_crate_items
  32:     0x7f8a8e5e8e19 - rustc_query_impl[66bcb677d88b0cff]::query_impl::hir_crate_items::invoke_provider_fn::__rust_begin_short_backtrace
  33:     0x7f8a8e5e5965 - rustc_query_impl[66bcb677d88b0cff]::execution::try_execute_query::<rustc_middle[48c27bf4c0e69198]::query::caches::SingleCache<rustc_middle[48c27bf4c0e69198]::query::erase::ErasedData<[u8; 8usize]>>, false>
  34:     0x7f8a8e2fc779 - rustc_query_impl[66bcb677d88b0cff]::query_impl::hir_crate_items::execute_query_non_incr::__rust_end_short_backtrace
  35:     0x7f8a8e2fd0de - rustc_interface[7c89e377fe0d0785]::passes::analysis
  36:     0x7f8a8e5e4549 - rustc_query_impl[66bcb677d88b0cff]::execution::try_execute_query::<rustc_middle[48c27bf4c0e69198]::query::caches::SingleCache<rustc_middle[48c27bf4c0e69198]::query::erase::ErasedData<[u8; 0usize]>>, false>
  37:     0x7f8a8e5e41a9 - rustc_query_impl[66bcb677d88b0cff]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  38:     0x7f8a8e5bdc50 - rustc_interface[7c89e377fe0d0785]::interface::run_compiler::<(), rustc_driver_impl[c394bdd49c3d550c]::run_compiler::{closure#0}>::{closure#1}
  39:     0x7f8a8e5a767e - std[f4db05f3516471ac]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[7c89e377fe0d0785]::util::run_in_thread_with_globals<rustc_interface[7c89e377fe0d0785]::util::run_in_thread_pool_with_globals<rustc_interface[7c89e377fe0d0785]::interface::run_compiler<(), rustc_driver_impl[c394bdd49c3d550c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  40:     0x7f8a8e5a7f20 - <std[f4db05f3516471ac]::thread::lifecycle::spawn_unchecked<rustc_interface[7c89e377fe0d0785]::util::run_in_thread_with_globals<rustc_interface[7c89e377fe0d0785]::util::run_in_thread_pool_with_globals<rustc_interface[7c89e377fe0d0785]::interface::run_compiler<(), rustc_driver_impl[c394bdd49c3d550c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[7428369f851be300]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  41:     0x7f8a8e5a8dac - <std[f4db05f3516471ac]::sys::thread::unix::Thread>::new::thread_start
  42:     0x7f8a880a597a - <unknown>
  43:     0x7f8a881292bc - <unknown>
  44:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.97.0-nightly (12f35ad39 2026-04-14) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [inherit_sig_for_delegation_item] inheriting delegation signature
#1 [fn_sig] computing function signature of `<impl at /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:3:1: 10:6>::bar`
#2 [lower_delayed_owner] lowering the delayed AST owner `<impl at /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:3:1: 10:6>::bar`
#3 [hir_crate_items] getting HIR crate items
#4 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 5 previous errors; 1 warning emitted

Some errors have detailed explanations: E0425, E0428, E0658.
For more information about an error, try `rustc --explain E0425`.

```

</p>
</details>

<!--
query stack:
#0 [inherit_sig_for_delegation_item] inheriting delegation signature
#1 [fn_sig] computing function signature of `<impl at /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:3:1: 10:6>::bar`
#2 [lower_delayed_owner] lowering the delayed AST owner `<impl at /tmp/icemaker_global_tempdir.hlcOhatqngCB/rustc_testrunner_tmpdir_reporting.QVJqXtOY9KAT/mvce.rs:3:1: 10:6>::bar`
#3 [hir_crate_items] getting HIR crate items
#4 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-min_generic_const_args +F-fn_delegation

--- a/tests/ui/delegation/inside-const-body-ice-155300.rs
+++ b/tests/ui/delegation/inside-const-body-ice-155300.rs
@@ -1,0 +1,18 @@
+#![feature(min_generic_const_args)]
+#![feature(fn_delegation)]
+
+pub struct S<const N: usize>;
+
+impl
+    S<
+        { //~ ERROR: complex const arguments must be placed inside of a `const` block
+            fn foo() {}
+            reuse foo::<> as bar;
+            reuse bar;
+            //~^ ERROR: the name `bar` is defined multiple times
+        },
+    >
+{
+}
+
+fn main() {}

--- a/tests/ui/delegation/inside-const-body-ice-155300.stderr
+++ b/tests/ui/delegation/inside-const-body-ice-155300.stderr
@@ -1,0 +1,24 @@
+error[E0428]: the name `bar` is defined multiple times
+  --> $DIR/inside-const-body-ice-155300.rs:11:13
+   |
+LL |             reuse foo::<> as bar;
+   |             --------------------- previous definition of the value `bar` here
+LL |             reuse bar;
+   |             ^^^^^^^^^^ `bar` redefined here
+   |
+   = note: `bar` must be defined only once in the value namespace of this block
+
+error: complex const arguments must be placed inside of a `const` block
+  --> $DIR/inside-const-body-ice-155300.rs:8:9
+   |
+LL | /         {
+LL | |             fn foo() {}
+LL | |             reuse foo::<> as bar;
+LL | |             reuse bar;
+LL | |
+LL | |         },
+   | |_________^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0428`.

